### PR TITLE
machine_core: Add Machine nspawn clone API

### PR DESCRIPTION
**Pull request by martinpitt**
There are tests which require two independent machines in terms of networking, services, packages, and file systems -- this can cover e.g. the shell's host switching, or cockpit-machine's VM migration or remote terminal functionality.

Previously we used destructive tests with `provision`ing multiple VMs; but this is expensive and prevents us from running these tests in tmt. But for the above tests which don't require kernel modules/functionality, we would be quite happy with the auxiliary machines being a container.

systemd-nspawn supports booting its own parent OS/file system with a transient overlay. Wrap that functionality into a convenient `Machine.clone_container()` API which returns a new Machine (NspawnMachine subclass) with the usual methods. Teach `SSHConnection` about a jump host (using `ProxyCommand`) to directly connect from the outside test to the inner nspawn container.

This first version only supports one cloned container at a time. Supporting multiple would be nice, but requires some reengineering of the SSHConnection master.

---

I tested this locally, and also sent a Cockpit PR to exercise this: https://github.com/cockpit-project/cockpit/pull/21105

### CODE REVIEW

**Comment by martinpitt** [comment]
So there are several problems with this -- [RHEL does not have networkd](https://cockpit-logs.us-east-1.linodeobjects.com/pull-21105-61d36c35-20241010-154738-centos-10-networking/log.html), and many/most OSes don't have a snapshot-capable file system, e.g. [debian-stable](https://cockpit-logs.us-east-1.linodeobjects.com/pull-21105-61d36c35-20241010-154726-debian-stable-networking/log.html). There, `systemd-nspawn --ephemeral` basically copies the entire file system and kills the VM. So we can't use that, but have to do something with `--overlay`, `--volatile`, etc. However,

```
# systemd-nspawn --machine=cockpit1 --volatile=overlay --network-veth --directory=/ --boot --bind /home
Spawning container on root directory is not supported. Consider using --ephemeral, --volatile=yes or --volatile=state.
```

`--volatile=yes` isn't supported properly by Debian (or any distro to date really), and `state` doesn't work either. The documentation of `--overlay` also says that this isn't supported with full OS.

**Comment by martinpitt** [comment]
Next step -- more legwork with manual overlays, which should work fine on ext4 now. I triggered another round of testing. Remaining big issue is that neither RHEL nor CoreOS have systemd-networkd, which makes configuring the inner and outer veth awkward. Need to find a way to do this with NM or other tools.

File: machine/machine_core/machine.py [hunk 1]
```diff
@@ -357,3 +358,93 @@ def curl(self, *args: str, headers: Mapping[str, str] = {}) -> str:
...
+                                        [ -z "$A" ] || break;
+                                        sleep 0.5;
+                                    done;
+                                    echo $A""", timeout=20).strip()
+
+        # wait until SSH is up, so that the ProxyCommand doesn't get stuck
+        parent.execute(f"until timeout 1 nc -z {address} 22; do sleep 1; done")
+
+        super().__init__(
+            address, parent.image, verbose, parent.label + "-" + label,
```
Comment (mvollmer) [inline machine/machine_core/machine.py:426]: Looks like `self.verbose` should be initialized earlier, in case the code above throws an error.

**Comment by martinpitt** [comment]
Some more ideas:

 - Unfortunately ssh [still cannot connect to a Unix socket](https://github.com/openssh/openssh-portable/pull/435). But we could play some tricks with socat to forward container port 22 to a host port via stdin/out or a unix socket?
 - Re-attempt to set up the veth pair manually with `ip` and a static IP, and figure out why the guest OS removes the IP. NM shouldn't touch veths.

**Comment by martinpitt** [comment]
This would have been really neat, but not being able to run this on RHEL sensibly is a dealbreaker. So I'm admitting defeat here.

**Pull request closed without merging**